### ISSUE TITLE
CMAKE BUILD_SHARED_LIBS flag produces empty DLL on windows

### ISSUE THREAD

**Issue by kvnnap** [syoyo/tinyobjloader#217]
I tried compiling a dynamic library on Windows using CMake and Visual Studio 2019.

I ran the following commands:

```bash
git clone https://github.com/syoyo/tinyobjloader
cd tinyobjloader
mkdir build
cd build
cmake .. -DBUILD_SHARED_LIBS:BOOL=ON
```

Which produced this output:

```bash
C:\Users\kvnna\source\repos\test\tinyobjloader\build>cmake .. -DBUILD_SHARED_LIBS:BOOL=ON
-- Building for: Visual Studio 16 2019
-- The C compiler identification is MSVC 19.22.27905.0
-- The CXX compiler identification is MSVC 19.22.27905.0
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.22.27905/bin/Hostx64/x64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.22.27905/bin/Hostx64/x64/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.22.27905/bin/Hostx64/x64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.22.27905/bin/Hostx64/x64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/kvnna/source/repos/test/tinyobjloader/build
```
I preceded to build using

```bash
cmake --build . --config Release
```

which produced the following output:

```bash
C:\Users\kvnna\source\repos\test\tinyobjloader\build>cmake --build . --config Release
Microsoft (R) Build Engine version 16.2.37902+b5aaefc9f for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

  Checking Build System
  Building Custom Rule C:/Users/kvnna/source/repos/test/tinyobjloader/CMakeLists.txt
  tiny_obj_loader.cc
  tinyobjloader.vcxproj -> C:\Users\kvnna\source\repos\test\tinyobjloader\build\Release\tinyobjloader.dll
  Building Custom Rule C:/Users/kvnna/source/repos/test/tinyobjloader/CMakeLists.txt
```

This results in a tinyobjloader.dll with no symbols exported, and hence empty (10kb in size as the compiler is probably removing all unreachable code).

I suspect the issue is that there is no `__declspec(dllexport)` attribute on the functions that need to be exported, and this seems to be required for the msvc compiler.

**Comment by syoyo** [comment]
tinyobjloader is a head-only library and typical use case is to embed it to an application, so it is not intended to use/compile as a standalone dll module.

Although, adding `__declspec(dllexport)` to API/classes would be trivial(with some cmake stuffs for dll build https://atomheartother.github.io/c++/2018/07/12/CPPDynLib.html ). If you need dll build, please send a PR.

**Comment by syoyo** [comment]
@kvnnap CMakeLists.txt is provided just for your reference. It's not recommended to use and not maintained well. 

> CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS

I didn't know that CMAKE feature.

https://stackoverflow.com/questions/54560832/cmake-windows-export-all-symbols-does-not-cover-global-variables

It looks promising.

**Comment by kvnnap** [comment]
@syoyo Thanks for your feedback. I will try the `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` solution, test it and submit a PR if all looks good.

**Comment by kvnnap** [comment]
It is worth noting that this fix applies to CMake 3.4 and above.
